crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -6118,12 +6125,30 @@ impl<'db> TypeInferenceBuilder<'db> {
                 todo_type!("ellipsis literal in type expression")
             }
 
-            // Other literals do not have meaningful values in the annotation expression context.
-            // However, we will we want to handle these differently when working with special forms,
-            // since (e.g.) `123` is not valid in an annotation expression but `Literal[123]` is.
-            ast::Expr::BytesLiteral(_literal) => todo_type!("bytes literal in type expression"),
-            ast::Expr::NumberLiteral(_literal) => todo_type!("number literal in type expression"),
-            ast::Expr::BooleanLiteral(_literal) => todo_type!("boolean literal in type expression"),
+            ast::Expr::BytesLiteral(_) => report_invalid_type_expression(format_args!(
+                "Bytes literals are not allowed in type expressions outside `typing.Literal[]` slices"
+            )),
```

Comment:
Should we include `typing.Annotated` here as `Annotated[int, b"hello"]` is valid? Similar comment for other messages.

Reply:
We could do. It feels sort-of like a rare edge-case, though? I think maybe we can revisit this error message when we have subdiagnostics and make it more precise with additional notes.

(Another nice improvement would be to have an autofix in your editor that imports `typing.Literal` and changes the annotation from `b"foo"` to `Literal[b"foo"]`, since that's what you _probably_ meant)

Reply:
Yeah, I think including all possible context would become a bit too verbose i.e., including both Literal and Annotated. We could probably just say "Invalid type expression" or "Bytes literals are not allowed as type expressions in this context".

That said, we can move forward with you have here and can iterate later.

Reply:
"in this context" sounds like a good compromise for now until we have subdiagnostics!